src/pm.rs

```diff
@@ -15,7 +16,53 @@ const SOURCE_PATH: &str = "/etc/apt/sources.list.d/atm.list";
 const STATE_PATH: &str = "/var/lib/atm/state";
 const STATE_DIR: &str = "/var/lib/atm/";
 const DPKG_STATE: &str = "/var/lib/dpkg/status";
-const MIRROR_URL: &str = "https://repo.aosc.io/debs";
+
+#[derive(Deserialize, Debug)]
+struct AptGenListStatus {
+    mirror: Vec<String>,
+}
+
+#[derive(Deserialize, Debug)]
+struct Mirror {
+    url: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct Mirrors {
+    mirror: HashMap<String, Mirror>,
+}
+
+pub fn get_mirror_url() -> Result<String> {
+    let mut status_file = fs::File::open("/var/lib/apt/gen/status.json")?;
+    let mut status_data = String::new();
+    status_file.read_to_string(&mut status_data)?;
+
+    let status_v: AptGenListStatus = serde_json::from_str(&status_data)?;
+    let mut mirror = String::from("origin");
+    if let Some(m) = &status_v.mirror.get(0) {
+        mirror = m.to_string();
+    }
+
+    let mut mirrors_file = fs::File::open("/usr/share/distro-repository-data/mirrors.yml")?;
+    let mut mirrors_data = String::new();
+    mirrors_file.read_to_string(&mut mirrors_data)?;
+
+    let mirrors_v: Mirrors = serde_yaml::from_str(&mirrors_data)?;
+    let mut mirror_url = String::from("https://repo.aosc.io/");
+    if let Some(m) = &mirrors_v.mirror.get(&mirror) {
+        mirror_url = m.url.to_string();
+    }
+
+    Ok(mirror_url)
+}
+
+lazy_static! {
+    static ref MIRROR_URL: String = format!(
+        "{}{}",
+        get_mirror_url().unwrap_or(String::from("https://repo.aosc.io/")),
+        "debs"
+    );
+}
```

Comment:
Is this still needed?

Reply:
There's a `format!` at where `MIRROR_URL` is used. If directly use the `format!` inside that `format!` it would raise `doesn't have a size known at compile-time` err.

Reply:
However, you have the option to integrate this `format!` into the other one by flattening the formatting string.

Reply:
On the second thought, now that you have duplicate `format!`'s in `main` and `pm`, you could consider exporting `MIRROR_URL`, so it will be run only once.